src/inmanta/resources.py

```diff
@@ -388,21 +518,61 @@ def __init__(self, _id: "Id") -> None:
 
         self.version = _id.version
 
-    def populate(self, fields: dict[str, Any], force_fields: bool = False) -> None:
+        self._references_model: dict[uuid.UUID, references.ReferenceModel] = {}
+        self._references: dict[uuid.UUID, references.Reference[references.RefValue]] = {}
+
+    def __getitem__(self, key: str) -> object:
+        """Support dict like access on the resource"""
+        if key in self.fields:
+            return getattr(self, key)
+
+        raise KeyError()
+
+    def __setitem__(self, key: str, value: object) -> None:
+        """Support dict like access on the resource. It is not possible to create new
+        attributes using setitem
+        """
+        if key in self.fields:
+            return setattr(self, key, value)
+
+        raise KeyError()
+
+    def get_reference_value(self, id: uuid.UUID) -> references.RefValue:
+        """Get a value of a reference"""
+        if id not in self._references:
+            if id not in self._references_model:
+                raise KeyError(f"The reference with id {id} is not defined in resource {self.id}")
+
+            model = self._references_model[id]
+            ref = references.reference.get_class(model.type).deserialize(model, self)
+            self._references[model.id] = ref
+
+        return self._references[id].get()
+
+    def resolve_all_references(self) -> None:
+        """Resolve all value references"""
+        for ref in self.references:  # type: ignore
+            model = references.ReferenceModel(**ref)
+            self._references_model[model.id] = model
+
+        for mutator in self.mutators:  # type: ignore
+            mutator = references.mutator.get_class(mutator["type"]).deserialize(references.MutatorModel(**mutator), self)
```

Comment:
We may want to define a base class resource to signal serialization failure , just to guide developer towards cleaner exceptions